### ISSUE TITLE
Regression: vlan_no_bridge default breaks hostapd dynamic VLANs joining bridges after 21.01.1

### ISSUE THREAD

**Issue by prbluebottle** [openwrt/openwrt#9944]
Platform: TP-Link Archer C7 v2 (but this is not platform-specific)
Versions: everything including the linked commit, i.e. every release including and after 21.02.2

I use dynamic VLANs with the full wpad-wolfssl package and this configuration stopped working after [this commit](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=62fa301cce9ede47e4e03433a0770e32bfb42a9d) made it into 21.02.2. If I effectively revert the change by hacking the `vlan_no_bridge` setting to 0, then my dynamic VLANs work again, i.e.
```
--- /lib/netifd/hostapd.sh      2022-05-25 22:47:01.000000000 +1200
+++ /rom/lib/netifd/hostapd.sh  2022-02-17 09:29:10.000000000 +1300
@@ -925,7 +925,7 @@
                if [ -n "$vlan_bridge" ]; then
                        append bss_conf "vlan_bridge=$vlan_bridge" "$N"
                else
-                       set_default vlan_no_bridge 0
+                       set_default vlan_no_bridge 1
                fi
                append bss_conf "vlan_no_bridge=$vlan_no_bridge" "$N"
                [ -n "$vlan_tagged_interface" ] && \
```
**Symptom**
The symptom of breakage is my dynamic VLANs do not get added to their respective bridges:
```
# brctl show
bridge name     bridge id               STP enabled     interfaces
br-untagged             7fff.a42bb0df3645       no              eth1.1
br-low          7fff.a42bb0df3645       no              eth1.12
br-guest                7fff.a42bb0df3645       no              eth1.11
br-management           7fff.a42bb0df3645       no              eth1.99
```
when this is the correct/desired result:
```
# brctl show
bridge name     bridge id               STP enabled     interfaces
br-untagged             7fff.a42bb0df3645       no              eth1.1
                                                        wlan5.1
br-low          7fff.a42bb0df3644       no              wlan5.12
                                                        eth1.12
br-guest                7fff.a42bb0df3645       no              wlan5.11
                                                        eth1.11
br-management           7fff.a42bb0df3645       no              eth1.99
```
I cannot fix this symptom with a supported configuration because the wlan5 interface does not exist when the bridge is created. Before tracking down the regression this was an even more hacky script that "fixed" it, run by `cron` every minute:
```
# cat hack-dynamic-vlan-interfaces-into-bridges.sh 
#!/bin/sh

/usr/sbin/brctl show br-low|/bin/grep wlan5 2>&1 > /dev/null
if [ "$?" -eq "1" ];then
        logger "HACK: Adding interface to br-low"
        /usr/sbin/brctl addif br-low wlan5.12
fi

/usr/sbin/brctl show br-untagged|/bin/grep wlan5 2>&1 > /dev/null
if [ "$?" -eq "1" ];then
        logger "HACK: Adding interface to br-untagged"
        /usr/sbin/brctl addif br-untagged wlan5.1
fi

/usr/sbin/brctl show br-guest|/bin/grep wlan5 2>&1 > /dev/null
if [ "$?" -eq "1" ];then
        logger "HACK: Adding interface to br-guest"
        /usr/sbin/brctl addif br-guest wlan5.11
fi

exit 0
```
**Configuration**
`/etc/config/network`
```
...
config device
        option name 'br-management'
        option type 'bridge'
        list ports 'eth1.99'

config device
        option name 'br-guest'
        option type 'bridge'
        list ports 'eth1.11'

config device
        option name 'br-untagged'
        option type 'bridge'
        list ports 'eth1.1'

config device
        option type 'bridge'
        option name 'br-low'
        list ports 'eth1.12'
...
```
`/etc/config/wireless`
```
...
config wifi-iface 'wifinet0'
        option device 'radio0'
        option mode 'ap'
        option isolate '1'
        option ifname 'wlan5'
        option auth_server '127.0.0.1'
        option auth_secret 'SNIP'
        option ieee80211k '1'
        option ieee80211r '1'
        option wnm_sleep_mode '1'
        option time_advertisement '2'
        option time_zone 'GMT0'
        option rrm_neighbor_report '1'
        option rrm_beacon_report '1'
        option bss_transition '1'
        option dynamic_vlan '2'
        option vlan_file '/etc/hostapd.vlan.wlan5'
        option eap_reauth_period '0'
        option encryption 'wpa2+ccmp'
        option ft_psk_generate_local '1'
        option mobility_domain 'SNIP'
        option ft_over_ds '0'
        option ssid 'SNIP'
```
`/etc/hostapd.vlan.wlan5`
```
1 wlan5.1 br-untagged
11 wlan5.11 br-guest
12 wlan5.12 br-low
```
I use `freeradius` for the EAP authentication but this is irrelevant as the bug occurs long before any authentication or clients even attempt to associate.

**Comment by memoz** [comment]
I encountered this problem too on 25.12.0-rc4 (Linksys WRT32X), but I think it's a side effect of migrating to DSA. My workaround is a bit hacky, but it works:
1. set option network 'lan' under wifi-device, so everything lands on br-lan.
```
root@ap1:~# brctl show
bridge name     bridge id               STP enabled     interfaces
br-lan          7fff.6038e0c1e088       no              lan1
                                                        lan2
                                                        lan3
                                                        lan4
                                                        phy0-ap0
                                                        phy1-ap0
                                                        wan
                                                        wifi10_r0
                                                        wifi1_r0
                                                        wifi1_r1
                                                        wifi3_r0
                                                        wifi5_r0
                                                        wifi8_r1
```
2. apk add ip-bridge
3. name dynamic vlan devices like this:
```
root@ap1:~# cat /etc/hostapd/default_radio0.vlan
1 wifi1_r0
3 wifi3_r0
5 wifi5_r0
10 wifi10_r0
root@ap1:~# cat /etc/hostapd/default_radio1.vlan
1 wifi1_r1
8 wifi8_r1
```
r is short for radio, so both APs can join the same vlan.
4. install this hotplug script courtesy of Google Gemini:
```
cat << 'EOF' > /etc/hotplug.d/net/99-mwlwifi-tagging
#!/bin/sh

# Only run when interface is ADDED
[ "$ACTION" = "add" ] || exit 0

# Match naming scheme: wifi<VID>_r<RADIO>
# Example: wifi3_r0 (VLAN 3 on Radio 0), wifi1_r1 (VLAN 1 on Radio 1)
case "$INTERFACE" in
    wifi*_r*)
        # 1. Extract VLAN ID
        # Remove prefix 'wifi' to get '3_r0'
        TEMP=${INTERFACE#wifi}
        # Remove suffix '_r*' to get '3'
        VID=${TEMP%%_r*}

        # Sanity Check: Ensure VID is a number
        if ! echo "$VID" | grep -qE '^[0-9]+$'; then
            logger -t wifi-vlan "Error: Could not parse VLAN ID from $INTERFACE"
            exit 0
        fi

        # Verify bridge exists
        if [ -d "/sys/class/net/br-lan" ]; then
            logger -t wifi-vlan "Detected dynamic interface $INTERFACE. Target VLAN: $VID"

            # 2. Enslave to Bridge
            if ip link set "$INTERFACE" master br-lan; then
                logger -t wifi-vlan "Enslaved $INTERFACE to br-lan"
            else
                logger -t wifi-vlan "Error: Failed to enslave $INTERFACE to br-lan"
                exit 1
            fi

            # 3. Configure VLAN PVID
            # We use 'bridge vlan' to set the PVID.
            # If successful, we log it.
            if bridge vlan add dev "$INTERFACE" vid "$VID" pvid untagged; then
                
                # If the target VLAN is NOT 1, we must remove the default VLAN 1 assignment
                # to prevent VLAN hopping/leaking.
                if [ "$VID" -ne 1 ]; then
                    bridge vlan del dev "$INTERFACE" vid 1 >/dev/null 2>&1
                    logger -t wifi-vlan "Configured PVID $VID for $INTERFACE (Removed default VLAN 1)"
                else
                    logger -t wifi-vlan "Configured PVID 1 for $INTERFACE (Default)"
                fi
            else
                logger -t wifi-vlan "Error: Failed to set VLAN $VID on $INTERFACE"
            fi

            # 4. Bring UP
            ip link set "$INTERFACE" up
            logger -t wifi-vlan "Interface $INTERFACE is now UP"
        else
            logger -t wifi-vlan "Error: br-lan not found. Cannot bridge $INTERFACE"
        fi
        ;;
esac
EOF
```
That's it. You'll see something like this if it works:
```
root@ap1:~# bridge vlan show
port              vlan-id
lan4              4 PVID Egress Untagged
lan3              4 PVID Egress Untagged
lan2              4 PVID Egress Untagged
lan1              4 PVID Egress Untagged
wan               1
                  3
                  5
                  8
                  10
br-lan            1
                  3
                  4
                  5
                  8
                  10
phy0-ap0          4 PVID Egress Untagged
phy1-ap0          4 PVID Egress Untagged
wifi10_r0         10 PVID Egress Untagged
wifi5_r0          5 PVID Egress Untagged
wifi3_r0          3 PVID Egress Untagged
wifi1_r0          1 PVID Egress Untagged
wifi8_r1          8 PVID Egress Untagged
wifi1_r1          1 PVID Egress Untagged

```
In summary, it's the same as adding vlan filtering in LuCI, but reads the config from vlan_file instead.
